beamer/health/notification_state.py

```diff
@@ -0,0 +1,34 @@
+import json
+import os
+from pathlib import Path
+
+
+class NotificationState:
+    cache_path: Path
+
+    def __init__(self, cache_path: Path) -> None:
+        self.cache_path = cache_path
+
+        try:
+            with open(os.path.join(self.cache_path, "breadcrumbs.json"), "r") as f:
+                self.notified_state = json.load(f)
+        except FileNotFoundError:
+            self.notified_state = {}
+
+    def is_set(self, request_id: str, notified_on: str) -> bool:
+        breadcrumb = self.notified_state.get(str(request_id))
+        if breadcrumb is None:
+            return False
+        else:
+            return breadcrumb.get(notified_on, False)
+
+    def update(self, request_id: str, notified_on: str) -> None:
+        breadcrumb = self.notified_state.get(request_id)
+        if breadcrumb is None:
+            self.notified_state[request_id] = {notified_on: True}
+        else:
+            breadcrumb[notified_on] = True
```

Comment:
If `breadcrumb` is just a `dict` that maps to a `bool`, consider instead using a `set` -- from this I imagine you only need membership testing, and that's what's `set` is made for.

Reply:
the problem is that for the same requestId -> I could have several notifications. I wouldn't want to notify the user for missed claim twice, but if the claim was missed, then made and then someone challenged - I sure want to have the challenge notification sent.

Reply:
> the problem is that for the same requestId -> I could have several notifications. I wouldn't want to notify the user for missed claim twice, but if the claim was missed, then made and then someone challenged - I sure want to have the 
challenge notification sent.

I do not see how using a `set` here prevents that. Could you please explain this in a bit more detail?

The way I understand, `update` just sets the value for the passed-in `notified_on` key to `True`:
```python
 def update(self, request_id: str, notified_on: str) -> None:
     breadcrumb = self.notified_state.get(request_id)
     if breadcrumb is None:
         self.notified_state[request_id] = {notified_on: True}
     else:
         breadcrumb[notified_on] = True
```

Functionally, this is completely equivalent to this set-based impl:
```python
 def update(self, request_id: str, notified_on: str) -> None:
     breadcrumb = self.notified_state.get(request_id)
     if breadcrumb is None:
         self.notified_state[request_id] = {notified_on}
     else:
         breadcrumb.add(notified_on)
```

I see no benefit of using a `dict` because you are not using the value part of the `dict` in any way: if an entry exists, the value is always `True`. And actually, that was precisely the pattern used in Python before sets were introduced; people just used a `dict` with the value typically always being `True` or `None`, i.e. a value that is simply ignored.

Reply:
> I decided that I won't got for a set here. I wrote the tests for the class and I was able to exchange dict for set, but it turns out that since I use json.dumps I can't use set as set is not JSON.serializable and I'll have to resort to pickle.

Yeah, that's unfortunate, but what is wrong with converting the set to a list just before serialization, and converting the list to a set after you deserialize JSON?

I imagine something like
```python
json.dump(list(self._notified_state), f)
```
and 
```python
self._notified_state = set(json.load(f))
```